evaluator: Check function call min and max arguments in new evaluator.

yanzhe-chen opened the pull request:
For github issue [#758](https://github.com/pingcap/tidb/issues/758).

1. Add min and max arguments checking.
2. Fix the incorrect min/max value for `builtinDate` function call.

ngaut commented:
LGTM @coocood PTAL

coocood commented:
Thank you!
And as always, please add some test case for this change.

coocood commented:
Make sure the test cases fails before this change and passes after this change.

yanzhe-chen commented:
@coocood Adding the test for this is a little bit tricky since we have to fake a wrong `builtin.Func` to trigger this failure. I am trying to do it and will update this request when finished.

yanzhe-chen commented:
@coocood I have added the test for this case. In order to create the fake `builtin.Func`, I have to import two extra modules (`expression/builtin` and `model`). I am not sure this is the optimal way. Please review it.

coocood commented on optimizer/evaluator/evaluator_test.go:
```diff
@@ -382,6 +384,30 @@ func (s *testEvaluatorSuite) TestConvert(c *C) {
+		Args:   []ast.ExprNode{},
+	}
+	_, err := Eval(ctx, expr)
+	c.Assert(err, NotNil)
```
You can assert equal error, like 
`c.Assert(ErrorInvalidOperation.Equals(err), IsTrue)`

coocood replied:
And add a valid call expression, assert the error is nil and value is expected.

coocood commented:
You don't need to create a fake `builtin.Funcs`, just use the existing one will do.

yanzhe-chen commented:
@coocood I have removed the fake `builtin.Funcs`. Also found a bug in handling unlimited upper bound of arguments. I fixed it and add related tests. PTAL.

yanzhe-chen commented:
@coocood Opps, I forgot to test the value for valid call expressions and use `c.Assert(ErrorInvalidOperation.Equals(err), IsTrue)` to assert equal error. I will refine that.

yanzhe-chen commented:
@coocood I have added the check for the return values from valid calls.

For assert equal error, I didn't found any use of `ErrorInvalidOperation.Equals` in the source code and the existing tests in `evaluator_test.go` also uses the `c.Assert(err, NotNil)` to assert errors.

So I think we'd better keep it consistent right now and replace all the old asserts to the new one in another independent commit.

Please help review it. Thanks!

coocood commented:
LGTM

It's great to see that you also have comment on test cases, many of our existing test cases are lacking comments.

The old test case was written when we didn't have `terror` package, so we can't easily compare errors.
But it's OK to just test error is nil for now, we have plan to refine old tests.

Thank you again. Looking forward for more PRs from you.
